File: cpp/examples/client_demo.cpp
#include <cstdio>
#include <vector>

#include "inferential/client.hpp"

int main() {
    auto conn = inferential::Connection("tcp://localhost:5555", "arm-01", "franka");
    auto policy = conn.model("manipulation-policy", 30.0f, 0);
    auto telemetry = conn.model("telemetry", 100.0f, 1);

    std::vector<float> joints(14, 0.1f);

    for (int step = 0; step < 5; ++step) {
        float urgency = std::min(1.0f, static_cast<float>(5 - step) / 5.0f);
        policy.observe()
            .urgency(urgency)
            .steps_remaining(static_cast<uint32_t>(5 - step))
            .tensor_f32("joint_positions", joints.data(), joints.size(), {14})
            .metadata("step", std::to_string(step))
            .send();

        std::printf("[step %d] observation sent\n", step);

        auto result = policy.get_result(100);
        if (result) {
            auto& output = *result;
            if (output.count("actions")) {
                auto [ptr, count] = output["actions"].as<float>();
                std::printf("[step %d] got %zu actions\n", step, count);
            }
        } else {
            std::printf("[step %d] timeout (no server)\n", step);
        }

        if (step % 2 == 0) {
            telemetry.observe()
                .urgency(0.1f)
                .tensor_f32("joint_positions", joints.data(), joints.size(), {14})
                .send();
        }
    }

    conn.close();
    std::printf("done\n");
    return 0;
}

File: rust/examples/client_demo.rs
use inferential::Connection;

fn main() {
    let conn = Connection::new("tcp://localhost:5555", "arm-01", "franka");
    let policy = conn.model("manipulation-policy", 30.0, 0);
    let telemetry = conn.model("telemetry", 100.0, 1);

    let joints: Vec<f32> = vec![0.1f32; 14];

    for step in 0..5u32 {
        let urgency = (1.0f32).min((5 - step) as f32 / 5.0);
        policy
            .observe()
            .urgency(urgency)
            .steps_remaining(5 - step)
            .tensor_f32("joint_positions", &joints, &[14])
            .metadata("step", &step.to_string())
            .send();

        println!("[step {}] observation sent", step);

        match policy.get_result(100) {
            Some(output) => {
                if let Some(actions) = output.get("actions") {
                    let values = actions.as_f32();
                    println!("[step {}] got {} actions", step, values.len());
                }
            }
            None => println!("[step {}] timeout (no server)", step),
        }

        if step % 2 == 0 {
            telemetry
                .observe()
                .urgency(0.1)
                .tensor_f32("joint_positions", &joints, &[14])
                .send();
        }
    }

    println!("done");
}
